core/consensus/src/adapter.rs

```diff
@@ -387,6 +444,178 @@ where
+        }
+
+        // check validators
+        for validator in block.header.validators.iter() {
+            if !authority_map.contains_key(&validator.address.as_bytes()) {
+                return Err(
+                    ConsensusError::VerifyBlockHeaderValidatorErr(block.header.height).into(),
+                );
+            } else {
+                let node = authority_map.get(&validator.address.as_bytes()).unwrap();
+
+                if node.vote_weight != validator.vote_weight
+                    || node.propose_weight != validator.vote_weight
+                {
+                    return Err(ConsensusError::VerifyBlockHeaderValidatorWeightErr(
+                        block.header.height,
+                    )
+                    .into());
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn verify_proof(
+        self: &Self,
+        ctx: Context,
+        block: &Block,
+        proof: &Proof,
+    ) -> ProtocolResult<()> {
+        // the block 0 has no proof, which is consensus-ed by community, not by chain
+
+        if block.header.height == 0 {
+            return Ok(());
+        };
+
+        if block.header.height != proof.height {
+            return Err(ConsensusError::VerifyBlockProofAndBlockHeightMismatchErr(
+                block.header.height,
+                proof.height,
+            )
+            .into());
+        }
+
+        let blockhash = Hash::digest(block.clone().encode_fixed()?);
+
+        if blockhash != proof.block_hash {
+            return Err(ConsensusError::VerifyBlockHashMismatchErr(block.header.height).into());
+        }
+
+        let metadata = self.get_metadata(
+            ctx,
+            block.header.state_root.clone(),
+            block.header.height,
+            block.header.timestamp,
+        )?;
+
+        let mut authority_list = metadata
+            .verifier_list
+            .iter()
+            .map(|v| Node {
+                address:        v.address.as_bytes(),
+                propose_weight: v.propose_weight,
+                vote_weight:    v.vote_weight,
+            })
+            .collect::<Vec<Node>>();
+
+        let vote = Vote {
+            height:     proof.height,
+            round:      proof.round,
+            vote_type:  VoteType::Precommit,
+            block_hash: proof.block_hash.as_bytes(),
+        };
+
+        let vote_hash = self.crypto.hash(protocol::Bytes::from(rlp::encode(&vote)));
+
+        let aggregated_signature = proof.signature.clone();
+
+        let signed_voters = extract_voters(&mut authority_list, &proof.bitmap).unwrap();
+
+        // check sig
+        match self.crypto.verify_aggregated_signature(
+            aggregated_signature,
+            vote_hash,
+            signed_voters.clone(),
+        ) {
+            Ok(()) => {}
+            Err(_) => {
+                return Err(ConsensusError::VerifyBlockProofErr(block.header.height).into());
+            }
+        };
```

Comment:
This should be extracted as an independent function like `fn verify_proof_signature()`.

Reply:
new function added